extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -1553,7 +1557,7 @@ private void handleQueryResults(
                                                      Tasks.DEFAULT_STORE_COMPACTION_STATE
                                                  );
 
-      if (!segments.isEmpty() && storeCompactionState) {
```

Comment:
Nit: Why don't we need to check for empty anymore?

Reply:
Earlier tombstone segments weren't annotated with `CompactionState`. Esp in case of empty content segments (but tombstones present), the `ShardSpec` wasn't available to do so. Now if there are no generated segments, we infer the ShardSpec from the `MSQControllerTask` spec and use that to annotate the state to tombstones.